packages/rum/src/domain/record/observer.ts

```diff
@@ -399,3 +424,24 @@ function initVisualViewportResizeObserver(cb: VisualViewportResizeCallback): Lis
     cancelThrottle()
   }
 }
+
+function initFrustrationObserver(lifeCycle: LifeCycle, frustrationCb: FrustrationCallback): ListenerHandler {
+  return lifeCycle.subscribe(LifeCycleEventType.RAW_RUM_EVENT_COLLECTED, (data) => {
+    if (
+      data.rawRumEvent.type === RumEventType.ACTION &&
+      data.rawRumEvent.action.type === ActionType.CLICK &&
+      data.rawRumEvent.action.frustration?.type &&
+      'events' in data.domainContext &&
+      data.domainContext.events
+    ) {
+      frustrationCb({
+        timestamp: data.rawRumEvent.date,
+        type: RecordType.FrustrationRecord,
+        data: {
+          frustrationTypes: data.rawRumEvent.action.frustration.type,
+          recordIds: data.domainContext.events.map((e) => getRecordIdForEvent(e)),
+        },
+      })
+    }
+  }).unsubscribe
+}
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** add a unit test for this function in `observer.spec.ts`